### PR TITLE
Full Matcher Function

### LINKED ISSUE #41: "Match girls to Workshops" button is broken
**Describe the bug**
Clicking the "Match girls to Workshops" button gives this error:
```
Script function not found: matchGirls For more information, see https://developers.google.com/apps-script/reference/base/menu#addItem(String,String)
```

This is because the `matchGirls` function was removed in a recent commit. It will need to be reintroduced or replaced with `main`.

**To Reproduce**
1. Open your development "Responses" sheet and update it to master.
2. Click on the "Match girls to Workshops" button in the "Great Explorations" menu.
3. See error

**Expected behavior**
Right now, anything that doesn't throw an error is fine. It might be good to fix this when we populate the spreadsheet.

### PR DESCRIPTION
**Pull request by Chris-Pyles**
**What did you do to resolve the issue? Explain your changes.**
This is a fully working version of the matching algorithm that produces viable matches for each student while filling each workshop to a set minimum capacity.

**Describe the steps to test your changes.**

-   Run main() from your dev file to run the matcher algorithm.
-   An error may appear due to randomness if a student cannot be given one of their preferred workshops during final matching.
-   Changing Matcher.minimumWorkshopFill to a value above 0.70 may result in an error where a workshop could not be sufficiently filled.
-   Changing Matcher.minimumWorkshopFill to a value too low (IE, 0.3) will result in an error where a student could not be given one of their preferred workshops during final matching.
-   The output sheet should display a complete set of matches for every student in the document, including each workshop's number, name, and location.

Resolves #41 #26 #36 #44 #33

### CODE REVIEW

File: Workshop.js [hunk 1]
```diff
@@ -33,16 +42,23 @@ function Workshop(name, number, capacity, location, sessionsPerWorkshop) {
...
+            this.slotsFilled += 1;
+            this.studentsAssigned.push(student);
+        }
+    }
+
+    this.hasReachedQuorum = function() {
+        return (this.slotsFilled >= this.minimumFill);
+    }
 
     this.toString = function() {
```
Comment (Izayuh) [inline Workshop.js:69]: I don't think we're using this function anymore.

File: Matcher.js [hunk 2]
```diff
@@ -20,8 +20,18 @@ function Matcher() {
         // An array containing all workshop objects sorted from least to most popular
         this.workshopsByPopularity = [];
```
Comment (Izayuh) [inline Matcher.js:20]: I would assume this variable name would sort from highest popularity to lowest, would it be better practice to rename this something like "workshopsByUnpopularity" or "workshopsByLowPopularity"?

File: Matcher.js [hunk 3]
```diff
@@ -98,44 +107,158 @@ function Matcher() {
...
+        Logger.log("Slots filled: " + workshop.slotsFilled + "/" + workshop.minimumFill);
+        Logger.log("Base Popularity: " + workshop.popularityScore);
+        throw new Error("Was unable to fill workshop " + workshop.name); // will handle this differently if it happens, EXTREMELY unlikely
+    }
+
+    /**
+     * The main matching algorithm, fills every workshop to its minimum with the students who prefer that workshop the most
+     */
+    this.reachMinimumForAll = function() {
+        this.fullEnough = [];
```
Comment (Izayuh) [inline Matcher.js]: We've mostly been using "Quorum" in the code to refer to workshops being "full enough", we may want to remain consistent with these variable names.

**Review by Izayuh — APPROVED** [review]

File: Student.js [hunk 4]
```diff
@@ -48,106 +48,42 @@ function Student(firstName, lastName, preferenceArray, grade, sessionsPerWorksho
...
-                // Then check the corresponding session.
-                // If there's still room in the session, we're good.
-                if (!session.hasReachedQuorum()) {
-                    return i;
-                }
+        if (this.fullyAssigned()) {
+            for (var i = 0; i < this.assignedWorkshops.length; i++) {
+                Logger.log(this.assignedWorkshops[i].name);
             }
+            Logger.log(workshop.name);
```
Comment (beszel) [inline Student.js]: Are these log messages still relevant? If they are, could you prepend each of them with a descriptive string to make the logs clearer?

**Review by beszel — CHANGES_REQUESTED** [review]

**Pull request merged**